### PR TITLE
[CP] Support Kimi K3 KDA CP

### PR DESCRIPTION
**Pull request by acisseJZhong**
## Summary

This PR replaces #4500 so the branch can participate in GitHub's native PR
stacks.

Enable context parallelism for Kimi K3's hybrid KDA/MLA decoder and
multimodal input path.

- KDA layers use Attention Gym's native context-parallel implementation. Each
  rank keeps its local sequence fragments while exchanging the short-convolution
  history and composing the recurrent state from preceding fragments.
- MLA layers can use all-gather K/V CP with contiguous or headtail partitions,
  or Ulysses CP with contiguous partitions.
- Multimodal preprocessing shards token-aligned vision-bank indices with the
  token IDs, allowing each rank to map replicated vision-encoder outputs to its
  local placeholders.

KDA and MLA reuse the same TorchTitan CP mesh. The all-gather configuration
supports contiguous and headtail partitions across the hybrid model; the
Ulysses configuration requires contiguous partitioning.

## Implementation

The integration follows Attention Gym's [delta-rule context-parallel example](https://github.com/meta-pytorch/attention-gym/blob/main/examples/linear/delta_rule_context_parallel.py).

The KDA path separates its local stages so the CP adapter can replace only the
distributed operations while sharing projection, normalization, gating, and
output logic with the non-CP path. Attention Gym provides:

- `context_parallel_conv_history` for convolution halo exchange; and
- `context_parallel_kda` for recurrent-state summary composition and the local
  KDA computation.

TorchTitan builds an Attention Gym routing plan that matches its contiguous or
headtail token assignment and passes the per-batch routing through every KDA
layer.

## Limitations

- KDA CP currently supports contiguous and headtail partitions; PTRR is not
  supported.
- Ulysses MLA requires contiguous partitioning; it does not support headtail.
- Current TorchTitan partitioning requires equal-sized fragments, where `T` is
  the token count and `C` is the CP degree:
  - contiguous: `T % C == 0`
  - headtail: `T % (2 * C) == 0`

  This is a limitation of the current PyTorch/TorchTitan equal-sized sharder,
  not Attention Gym; Attention Gym routing accepts arbitrary fragment ranges.
- Kimi K3 sample packing is not yet supported. Routing currently models the
  batch as one global sequence.

## Tests

Tested on NVIDIA H100 GPUs with Attention Gym main at commit `7418529`.

Targeted unit tests:

```text
pytest -q tests/unit_tests/gpu/test_kda_attention.py \
  tests/unit_tests/gpu/test_kimi_k3.py
# 13 passed

pytest -q tests/unit_tests/cpu/test_integration_test_definitions.py
# 12 passed
```

Two-GPU integration tests (10 training steps each):

```text
python -m tests.integration_tests.run_tests <output-dir> \
  --test_suite b200 \
  --test_name kimi_k3_mm_allgather_kv_cp \
  --ngpu 2 --no-parallel
# rc=0

python -m tests.integration_tests.run_tests <output-dir> \
  --test_suite b200 \
  --test_name kimi_k3_mm_ulysses_cp \
  --ngpu 2 --no-parallel
# rc=0
```

Deterministic BF16 runs compared parent `15a45326f` at CP=1 with `35f2a6601`
at CP=2, using the same seed checkpoint (`seed=42`), 256 tokens per step, and
100 training steps. CP=2 used Attention Gym CP for KDA and either headtail
all-gather K/V or contiguous Ulysses for MLA. Values are measured at steps 1,
10, and 100; each diff is the absolute percentage difference from CP=1.
The parent baseline was run on H100 with only its Blackwell capability guard
relaxed; its computation was unchanged.

| Step | CP=1 loss | All-gather loss | Diff | Ulysses loss | Diff |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | `12.624805` | `12.616027` | `0.070%` | `12.627501` | `0.021%` |
| 10 | `4.009669` | `4.218186` | `5.200%` | `4.057774` | `1.200%` |
| 100 | `2.942050` | `2.968832` | `0.910%` | `2.862944` | `2.689%` |

| Step | CP=1 grad norm | All-gather grad norm | Diff | Ulysses grad norm | Diff |
| ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | `25.625` | `25.625` | `0.000%` | `25.75` | `0.488%` |
| 10 | `6.5` | `8.625` | `32.692%` | `5.75` | `11.538%` |
| 100 | `5.5625` | `5.5625` | `0.000%` | `6.0` | `7.865%` |

### CODE REVIEW

**Comment by acisseJZhong** [comment]
This PR replaces #4500; the earlier review history remains available there.

File: torchtitan/config/transform/context_parallel.py [hunk 1]
```diff
@@ -46,3 +46,21 @@ def transform(self, model: Module.Config) -> Module.Config:
                 attention.inner_attention, self.inner_attention
             )
         return model
+
+
+@dataclass(kw_only=True, slots=True)
+class KDAContextParallelTransform(ModelConfigTransform):
```
Comment (tianyu-l) [inline torchtitan/config/transform/context_parallel.py]: Not clear why we should keep this separate from `ContextParallelTransform`
- it makes that name ambiguous
- what if I have a hybrid model that applies one but not the other?

File: torchtitan/models/kimi_k3/cp_kda.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                raise ValueError(
+                    "KDA context parallelism requires VarlenMetadata in "
+                    "attention_masks or attention_masks['kda']."
+                )
+            cu_seqlens_global = kda_metadata.cu_seq_q.tolist()
+
+        # Kimi K3 validation rejects PTRR, so a non-None balancer is head-tail.
+        input_dict["kda_cp_routing"] = _build_kda_context_parallel_routing(
+            cu_seqlens_global=cu_seqlens_global,
+            conv_kernel_size=backend_config.conv_kernel_size,
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/cp_kda.py]: this is confusing -- you have `conv_kernel_size` in the config
- why making this a class method? If it's instance method you'd easily access the field
- why this method belongs to Attn not Attn.Config?

File: torchtitan/models/kimi_k3/model.py [hunk 3]
```diff
@@ -237,6 +244,7 @@ def forward(
         block_residual_TND: torch.Tensor,
         attention_masks: KimiK3AttentionMaskDict | None = None,
         positions: torch.Tensor | None = None,
+        kda_cp_routing: ContextParallelRouting | None = None,
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/model.py:250]: Instead of keep adding things to the list, sounds that we should enhance `attention_masks` to be a general holder of such info, e.g. `context_metadata`.

File: torchtitan/models/kimi_k3/kda.py [hunk 4]
```diff
@@ -142,7 +165,41 @@ def forward(
         dt_bias_HK: torch.Tensor,
         *,
         cu_seqlens: torch.Tensor | None,
+        routing: ContextParallelRouting | None,
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/kda.py:169]: add docstring at least for public methods

File: torchtitan/config/transform/context_parallel.py
Comment (fegin) [inline torchtitan/config/transform/context_parallel.py]: This PR should address this TODO, then you don't need `KDAContextParallelTransform`. Concretely, you can change inner_attention from `type[Module]` to `Dict[Config, type[Module]]`. So you can map different non-CP inner attentions to different CP inner attentions. This design doesn't address the selective gather problem, which the inner attentions are the same but the mask is different. So this is not a final form. We can think about the correct "key" should be.

File: torchtitan/models/kimi_k3/cp_kda.py (hunk 2 above)
Comment (acisseJZhong) [reply torchtitan/models/kimi_k3/cp_kda.py]: make cp_shard_metadata a method of Config. This 
- makes decoder logic simpler as we don't need to find owner class and call `owner.cp_shard_metadata`, we just call `config.cp_shard_metadata` 
- no longer need to pass `backend_config`.

File: torchtitan/models/kimi_k3/model.py (hunk 3 above)
Comment (acisseJZhong) [reply torchtitan/models/kimi_k3/model.py]: created `KimiK3AttentionMetadata` for this

File: torchtitan/config/transform/context_parallel.py [hunk 5]
```diff
@@ -20,29 +20,47 @@
...
+    instance, such as its FQN and mask key. Multiple attention instances can
+    share one local config type while only some of their masks require K/V
+    all-gather; the current mapping assigns all such instances the same CP
+    backend.
     """
 
-    inner_attention: type[Module]
-    """Replacement inner attention; must inherit ``CPInnerAttention``."""
+    inner_attention: dict[type[Module.Config], type[Module]]
+    """Map each local config type to its CP backend implementation."""
```
Comment (fegin) [inline torchtitan/config/transform/context_parallel.py:33]: @jinsooihm this is what we discussed. It's close to LoRA. But with sliding window attention, this is not enough.

Comment (jinsooihm) [reply torchtitan/config/transform/context_parallel.py:33]: I see. I guess a general version could be a configurable function mapping Config to Module?

Comment (fegin) [reply torchtitan/config/transform/context_parallel.py:33]: Yes, maybe a better way is a Callable.
